File: src/util/date.ts
/**
 * Converts a date string into a JS Date object while accounting for time zone
 * offset to make sure that the date returned matches the string given.
 * 
 * @param dateString Date as YYYY-MM-DD
 * @returns Corresponding JS Date at midnight local time
 */
export function getNormalizedDate(dateString: string) {
    const buffer = new Date(dateString + 'T00:00:00Z')
    buffer.setMinutes(buffer.getMinutes() + buffer.getTimezoneOffset())
    return buffer
}

/**
 * @param date JS Date to convert. If not specified, the current date will be
 * used instead.
 * @returns Representation of date as YYYY-MM-DD
 */
export function getDateString(date?: Date) {
    if (date === undefined) {
        date = new Date()
    }
    return date.toISOString().split('T')[0] as string
}

/**
 * Extract a list of month and week date ranges from the given base date range.
 * 
 * @param dateStart The first date to consider (inclusive).
 * @param dateEnd The last date to consider (inclusive).
 * @returns Object containing months and weeks arrays with date range strings.
 */
export function getMonthsAndWeeksFromDateRange(dateStart: string, dateEnd: string) {
    const months = [] as { dateStart: string, dateEnd: string }[]
    const weeks = [] as { dateStart: string, dateEnd: string }[]
    
    // Iterate our range and identify weeks and months within it
    const maxDate = getNormalizedDate(dateEnd)
    const bufferDate = getNormalizedDate(dateStart)
    let weekStart = getDateString(bufferDate)
    let monthStart = getDateString(bufferDate)
    while (bufferDate <= maxDate) {
        const dateString = getDateString(bufferDate)
        const tomorrow = getNormalizedDate(dateString)
        tomorrow.setDate(tomorrow.getDate() + 1)
        const tomorrowString = getDateString(tomorrow)

        // Finish a week?
        if (bufferDate.getDay() === 6) {
            weeks.push({
                dateStart: weekStart,
                dateEnd: dateString
            })
            weekStart = tomorrowString
        }

        // Finish a month?
        if (tomorrow.getDate() === 1) {
            months.push({
                dateStart: monthStart,
                dateEnd: dateString
            })
            monthStart = tomorrowString
        }

        // Prep the next iteration
        bufferDate.setDate(bufferDate.getDate() + 1)
    }

    // Finish out any incomplete weeks or months
    bufferDate.setDate(bufferDate.getDate() - 1)
    const dateString = getDateString(bufferDate)
    if (dateString >= weekStart) {
        weeks.push({
            dateStart: weekStart,
            dateEnd: dateString
        })
    }
    if (dateString >= monthStart) {
        months.push({
            dateStart: monthStart,
            dateEnd: dateString
        })
    }

    return {
        months,
        weeks
    }
}

File: src/wasm/util/date.cpp
#include <ctime>
#include <iomanip>
#include <sstream>

#include "date.h"

using namespace std;

time_t getNormalizedTimestamp(string dateString) {
    tm buffer = {};
    istringstream ss(dateString);
    ss >> get_time(&buffer, "%Y-%m-%d");
    if (ss.fail()) {
        return -1;
    }
    
    return mktime(&buffer);
}

string getDateString(time_t timestamp) {
    if (timestamp == 0) {
        time(&timestamp);
    }

    tm* buffer = gmtime(&timestamp);
    ostringstream ss;
    ss << put_time(buffer, "%Y-%m-%d");

    return ss.str();
}

MonthAndWeekRanges getMonthsAndWeeksFromDateRange(string dateStart, string dateEnd) {
    MonthAndWeekRanges results;

    // Iterate our range and identify weeks and months within it
    const time_t maxDate = getNormalizedTimestamp(dateEnd);
    time_t bufferTs = getNormalizedTimestamp(dateStart);
    string weekStart = getDateString(bufferTs);
    string monthStart = weekStart;
    while (bufferTs <= maxDate) {
        const tm* bufferDate = gmtime(&bufferTs);
        const string dateString = getDateString(bufferTs);
        const time_t tomorrowTs = bufferTs + (60 * 60 * 24);
        const tm* tomorrowDate = gmtime(&tomorrowTs);
        const string tomorrowString = getDateString(tomorrowTs);

        // Finish a week?
        if (bufferDate->tm_wday == 6) {
            DateRange week;
            week.dateStart = weekStart;
            week.dateEnd = dateString;
            results.weeks.push_back(week);

            weekStart = tomorrowString;
        }

        // Finish a month?
        if (tomorrowDate->tm_mday == 1) {
            DateRange month;
            month.dateStart = monthStart;
            month.dateEnd = dateString;
            results.months.push_back(month);

            monthStart = tomorrowString;
        }

        // Prep the next iteration
        bufferTs = tomorrowTs;
    }

    // Finish out any incomplete weeks or months
    bufferTs -= (60 * 60 * 24);
    const string dateString = getDateString(bufferTs);
    if (dateString >= weekStart) {
        DateRange week;
        week.dateStart = weekStart;
        week.dateEnd = dateString;
        results.weeks.push_back(week);
    }
    if (dateString >= monthStart) {
        DateRange month;
        month.dateStart = monthStart;
        month.dateEnd = dateString;
        results.months.push_back(month);
    }

    return results;
}
